plenum/test/simulation/sim_network.py

```diff
@@ -1,14 +1,100 @@
 from collections import OrderedDict
+from contextlib import contextmanager
 from functools import partial
 from logging import getLogger
-from typing import Any, Iterable, Optional, Callable
+from typing import Any, Iterable, Optional, Callable, NamedTuple, Union, List, Tuple
 
 from plenum.common.event_bus import ExternalBus
 from plenum.common.timer import TimerService
 from plenum.server.consensus.utils import replica_name_to_node_name
 from plenum.test.simulation.sim_random import SimRandom
 
 
+# TODO: INDY-2324 Should we namespace those definitions somehow?
+Discard = NamedTuple('Discard', [('probability', float)])
+Deliver = NamedTuple('Deliver', [('min_delay', float),
+                                 ('max_delay', float)])
+Stash = NamedTuple('Stash', [])
+Action = Union[Discard, Deliver, Stash]
+
+Selector = Callable[[Any, str, str], bool]  # predicate receiving message and names of source and destination peers
+Processor = NamedTuple('Processor', [('action', Action),
+                                     ('selectors', Iterable[Selector])])
+
+
+def message_frm(frm: Union[str, Iterable[str]]) -> Selector:
+    def _selector(_msg: Any, _frm: str, _dst: str):
+        return _frm == frm if isinstance(frm, str) else _frm in frm
+    return _selector
+
+
+def message_dst(dst: Union[str, Iterable[str]]) -> Selector:
+    def _selector(_msg: Any, _frm: str, _dst: str):
+        return _dst == dst if isinstance(dst, str) else _dst in dst
+    return _selector
+
+
+def message_type(t: Union[type, Iterable[type]]) -> Selector:
+    def _selector(_msg: Any, _frm: str, _dst: str):
+        return isinstance(_msg, t) if isinstance(t, type) else isinstance(_msg, tuple(t))
+    return _selector
+
+
+class ProcessingChain:
+    ScheduleDelivery = Callable[[float, Any, str, str], None]  # callable receiving delay, msg, source and destination peers
+
+    def __init__(self, random: SimRandom, schedule_delivery: ScheduleDelivery):
+        self._random = random
+        self._schedule_delivery = schedule_delivery
+        self._processors = []  # type: List[Processor]
+        self._stashed_messages = []  # type: List[Tuple[Any, str, str]]
+        self._default_min_latency = 0.01
+        self._default_max_latency = 0.5
+
+    def set_default_latency(self, min_value: float, max_value: float):
+        self._default_min_latency = min_value
+        self._default_max_latency = max_value
+
+    def add(self, p: Processor):
+        self._processors.append(p)
+
+    def remove(self, p: Processor):
+        if p not in self._processors:
+            return
+        self._processors.remove(p)
+        self._pump_messages()
+
+    def process(self, msg: Any, frm: str, dst: str):
+        for p in reversed(self._processors):
+            if not all(selector(msg, frm, dst) for selector in p.selectors):
+                continue
+
+            if isinstance(p.action, Discard):
```

Comment:
It may make sense to follow chain-of-responsibility pattern more formally (have a Handler interface so that every processor implements it) to avoid `isinstance` checks.

Reply:
I agree that having `isinstance` checks is often an anti-pattern, however implementing full chain-of-responsibility looked like an overkill for this simple use case.